### PR TITLE
lib: gui: lvgl: update to v7.0.2

### PR DESCRIPTION
**Pull request by gmarull**
Update LVGL to version 7.0.2. Notable changes in v7 include:

- New drawing system (note that it uses much more ROM than previous
  versions)
- New style system
- Some objects have been renamed (current changes do not align yet
  Zephyr Kconfig settings with LVGL)
- New fonts with more sizes (e.g. Montserrat, replacing Roboto)
- Theme changes (most have been removed, default is now Material)

Note that constant defaults have been aligned with LVGL.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

### CODE REVIEW

**Comment by gmarull** [comment]
This PR partially supersedes #26654 (only update to 7.0.2).

NOTE: ROM optimization is being analyzed by upstream maintainers: https://github.com/lvgl/lvgl/issues/1669

File: lib/gui/lvgl/lv_conf.h
Comment (pabigot) [inline lib/gui/lvgl/lv_conf.h:137]: This could be made more compact and understandable as:

    #define LV_USE_BLEND_MODES IS_ENABLED(CONFIG_LVGL_BLEND_MODES)

Multiple instances.  I won't insist on it because existing code would have to change and this file is horrible anyway.

File: lib/gui/lvgl/lv_conf.h
Comment (pabigot) [inline lib/gui/lvgl/lv_conf.h:347]: This is probably OK but there are situations where Kconfig symbols are undefined, so unless this is guaranteed to get defined it should be:

    #if (CONFIG_LVGL_THEME_CUSTOM - 0)

**Review by pabigot — DISMISSED** [review]
I haven't tested any of this.

The [documented agreement](https://github.com/zephyrproject-rtos/zephyr/pull/26654#issuecomment-659503003) is that font settings defaults should not change, but I'm not sure that's accurate.  I'm OK with aligning all the defaults with the update to a new major version.  We may need to discuss this again to clarify.

I [still want](https://github.com/zephyrproject-rtos/zephyr/pull/26654#pullrequestreview-449847671) documentation of the impact this has on existing applications in terms of things that silently changed.  I.e. a summary in `doc/releases/release-notes-2.4.rst` and a pointer to more details of what specifically changed in Kconfig names and default values somewhere else, probably `lib/gui/lvgl/README.rst`.

File: lib/gui/lvgl/lv_conf.h
Comment (gmarull) [reply lib/gui/lvgl/lv_conf.h:137]: Thanks for the suggestion @pabigot. I think that can be addressed in the planned PR where re-organization of Kconfig will be handled.

File: lib/gui/lvgl/lv_conf.h
Comment (gmarull) [reply lib/gui/lvgl/lv_conf.h:347]: Theme selection is a choice now, so at least 1 theme will be enabled.

File: lib/gui/lvgl/Kconfig
Comment (gmarull) [reply lib/gui/lvgl/Kconfig:45]: I think it was agreed to keep enabled options to a minimum (as is now), but align with LVGL for the rest (constants and names). I can do constant alignment in the planned PR that will come after this one if it's preferred.

**Comment by gmarull** [comment]
> I haven't tested any of this.
> 
> The [documented agreement](https://github.com/zephyrproject-rtos/zephyr/pull/26654#issuecomment-659503003) is that font settings defaults should not change, but I'm not sure that's accurate. I'm OK with aligning all the defaults with the update to a new major version. We may need to discuss this again to clarify.

Note that default font has not changed (UNSCII 8px). LVGL7 has new fonts/sizes (no Roboto anymore) and relates fonts with themes in a different manner.

> 
> I [still want](https://github.com/zephyrproject-rtos/zephyr/pull/26654#pullrequestreview-449847671) documentation of the impact this has on existing applications in terms of things that silently changed. I.e. a summary in `doc/releases/release-notes-2.4.rst` and a pointer to more details of what specifically changed in Kconfig names and default values somewhere else, probably `lib/gui/lvgl/README.rst`.

I'll add some notes.

File: lib/gui/lvgl/Kconfig
Comment (vanwinkeljan) [reply lib/gui/lvgl/Kconfig:45]: No problem for me keep it as is.

File: samples/display/lvgl/boards/native_posix.conf
Comment (vanwinkeljan) [inline samples/display/lvgl/boards/native_posix.conf:7]: These changes should not be needed if https://github.com/zephyrproject-rtos/zephyr/pull/26827 gets merged (same for native_posix_64 and hsdk)

File: doc/releases/release-notes-2.4.rst
Comment (gmarull) [reply doc/releases/release-notes-2.4.rst]: fixed

File: doc/releases/release-notes-2.4.rst
Comment (gmarull) [reply doc/releases/release-notes-2.4.rst]: fixed, thanks

**Review by pabigot — COMMENTED** [review]
Documentation is ok for now.  Because I'm not testing this I'm not going to approve it at this time; I could do some testing on a couple platforms later if necessary.

**Review by MaureenHelm — APPROVED** [review]
Tested on `mimxrt1050_evk`

File: samples/display/lvgl/boards/native_posix.conf
Comment (gmarull) [reply samples/display/lvgl/boards/native_posix.conf:7]: @vanwinkeljan while it now works without these settings, the allocated memory is now higher than the needed amount. I think it is still better to specify such settings for `CONFIG_LVGL_BUFFER_ALLOC_STATIC` so that no memory is wasted.

Comment (vanwinkeljan) [reply samples/display/lvgl/boards/native_posix.conf:7]: Memory will be not a problem on native_posix so no need to reduce the buffer on native_posix.
It could be an argument to lineup de SDL display resolution with the LVGL defaults.

**Comment by vanwinkeljan** [comment]
Tested on `nrf52840dk_nrf52840` with `adafruit_2_8_tft_touch_v2`
Tested on `reel_board` v1

File: lib/gui/lvgl/Kconfig.themes
Comment (vanwinkeljan) [inline lib/gui/lvgl/Kconfig.themes:9]: Better use `LVGL_THEME_EMPTY` as default as it reduces the rom footprint with 3.7k.
Tested the sample on the `nrf` board and it is still OK

Comment (gmarull) [reply lib/gui/lvgl/Kconfig.themes:9]: Empty theme doesn't look really good unless you apply some custom style (it's actually the purpose). So if `EMPTY` is the default I think sample should be adjusted. Note that there is also a gain by disabling logging (now enabled by default).

image

File: samples/display/lvgl/boards/native_posix.conf
Comment (gmarull) [reply samples/display/lvgl/boards/native_posix.conf:7]: Even if memory is not a problem on posix, it is still a waste of memory for nothing (it actually doubles to be precise, +144Kb).

File: lib/gui/lvgl/Kconfig.themes
Comment (vanwinkeljan) [reply lib/gui/lvgl/Kconfig.themes:9]: I have no problem if `CONFIG_LVGL_THEME_MATERIAL` is added to the sample `prj.conf` file. But I like the idea of setting the defaults in such a way that you get a minimal build and that as a user you need to enable what you need to get your app working.

Comment (gmarull) [reply lib/gui/lvgl/Kconfig.themes]: Ok, makes sense. Default is now empty but sample enables material.

**Comment by gmarull** [comment]
Tested with [this ILI9341 based display](https://www.buydisplay.com/2-8-inch-tft-touch-shield-for-arduino-w-capacitive-touch-screen-module)

**Review by vanwinkeljan — APPROVED** [review]

**Comment by jfischer-phytec-iot** [comment]
ST7789V on st7789v_waveshare_240x240 shows wrong color with this PR, but works fine with the version on master, #27087 fixes that.

**Comment by jfischer-phytec-iot** [comment]
@carlescufi @pabigot @vanwinkeljan @MaureenHelm does everyone agree to merge it despite the problem with the flash consumption?

**Comment by pabigot** [comment]
I'm OK with merging it as is if we also open a medium priority issue targeting 2.4 that lvgl be updated again when upstream has provided a newer 7.x release that improves the memory use.  I don't want this to be forgotten.

**Comment by gmarull** [comment]
@pabigot 

> I'm OK with merging it as is if we also open a medium priority issue targeting 2.4 that lvgl be updated again when upstream has provided a newer 7.x release that improves the memory use. I don't want this to be forgotten.

https://github.com/zephyrproject-rtos/zephyr/issues/27095

**Comment by carlescufi** [comment]
> @carlescufi @pabigot @vanwinkeljan @MaureenHelm does everyone agree to merge it despite the problem with the flash consumption?

I agree to merge now that I have seen that LVGL themselves are taking this seriously and addressing this:
https://github.com/lvgl/lvgl/issues/1669

Given that @pabigot and @vanwinkeljan seems to also agree, I will merge this after I get an ack from you @jfischer-phytec-iot .

**Comment by vanwinkeljan** [comment]
Just had a  look at the diff between `lv_conf_template.h` in our lvgl v7 fork on the one the lvgl team is using for size reduction and found missing flags missing in our lv_conf.h. Need to have a further look at this hence DNM

**Comment by gmarull** [comment]
@vanwinkeljan is there any missing flag from v7.0.2 release?

**Comment by vanwinkeljan** [comment]
> is there any missing flag from v7.0.2 release?

Could not find following v7.0.2 flag in the update lv_conf.h
```#define LV_USE_GPU_STM32_DMA2D  0```

This one possibly does no harm but would like to double check if there are others

**Comment by embeddedt** [comment]
I'm pretty sure `LV_USE_GPU_STM32_DMA2D` will only affect rendering performance. There should be little to no impact on code size from that one.

**Comment by vanwinkeljan** [comment]
ok looks like no other flags are missing

**Review by vanwinkeljan — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**
